File: 09/C++/matrix.cpp
/*
Programa:   Matrix
Arquivo:    matrix.cpp
*/
#include <iostream>

using namespace std;

class Agente
{
    private:
        string nome;
        bool modo_agente = false;
        bool get_mode(){
            return this->modo_agente;
        }
    protected:
        string profissao;
    public:
        void set_nome(string nome)
        {
            this->nome = nome;
        }
        void mode_on() {
            this->modo_agente = true;
        }
        void apresentacao()
        {
            if(this->get_mode()){
                cout << "AGENTE SMITH !!!!" << endl;
            } else {
                cout << "\n Olá ! Meu nome é " << this->nome << endl;
            }
        }
//        virtual void set_profissao() = 0;
};
class Empresario: public Agente {
    private:
        string empresa;
    public:
        Empresario(string nome, string empresa) {
            set_nome(nome);
            this->empresa = empresa;
//            this->set_profissao();
        }
        void apresentacao()
        {
            Agente::apresentacao();
            cout << "sou executivo da " << this->empresa << endl;
        }
//        void set_profissao(){   this->profissao = "Empresário";    }
};
class Professor : public Agente
{
    private:
        string escola;
    public:
        Professor(string nome, string escola) {
             set_nome(nome);
            this->escola = escola;
//            this->set_profissao();
        }
        void apresentacao()
        {
            Agente::apresentacao();
            cout << "Leciono na escola " << this->escola << endl;
        }
//        void set_profissao(){   this->profissao = "Professor";     }
} ;
class Advogado : public Agente
{
    private:
        string OAB;
    public:
        Advogado(string nome, string OAB) {
            set_nome(nome);
            this->OAB = OAB;
//            this->set_profissao();
        }
        void apresentacao()
        {
            Agente::apresentacao();
            cout << "sou advogado e meu OAB é: " << this->OAB << endl;
        }
//        void set_profissao(){ this->profissao = "Advogado";        }
} ;

File: 09/Python3/matrix.py
#!/usr/bin/python3
# coding: utf-8
""" Módulo Matrix. """
import abc


class Agente(metaclass=abc.ABCMeta):
    """ Classe abstrata Agente. """
    modo_agente = False

    def mode_on(self):
        """ Método para ativação do agente. """
        self.modo_agente = True

    @abc.abstractmethod
    def __init__(self, nome):
        """ Construtor abstrado da classe base. """
        self.nome = nome

    @abc.abstractmethod
    def apresentacao(self):
        """ Método de Apresentacao. """
        if self.modo_agente:
            print ('Oi, sou o Agente Smith')
        else:
            print ('Olá ! Meu nome é ' + str(self.nome))


class Executivo(Agente):
    """ Classe Executivo. """
    def __init__(self, nome, empresa):
        Agente.__init__(self, nome)
        self.empresa = empresa
        self.profissao = "Executivo"

    def apresentacao(self):
        Agente.apresentacao(self)
        print ('Sou executivo(a) na empresa ' + str(self.empresa))


class Professor(Agente):
    """ Classe Professor. """
    def __init__(self, nome, escola):
        Agente.__init__(self, nome)
        self.escola = escola
        self.profissao = "Professor"

    def apresentacao(self):
        Agente.apresentacao(self)
        print ('Sou professor(a) na escola ' + str(self.escola))


class Advogado(Agente):
    """ Classe Advogado. """
    def __init__(self, nome, oab):
        Agente.__init__(self, nome)
        self.oab = oab
        self.profissao = "Advogado"

    def apresentacao(self):
        Agente.apresentacao(self)
        print ('Sou advogado(a). Meu OAB: ' + str(self.oab))

